Spark runtime

cmadam opened the pull request:
## Why are these changes needed?
Add Spark to the data prep kit

## Related issue number (if any).
https://github.com/IBM/data-prep-kit/issues/85

daw3rd committed:
initial spark library directory

Signed-off-by: David Wood <dawood@us.ibm.com>

daw3rd committed:
fix make venv in new spark lib

Signed-off-by: David Wood <dawood@us.ibm.com>

daw3rd committed:
spark launcher test running with spark stubbed out

Signed-off-by: David Wood <dawood@us.ibm.com>

daw3rd committed:
fix spark pyproject.toml to use spark wheel name

Signed-off-by: David Wood <dawood@us.ibm.com>

cmadam committed:
First running version of noop local in Spark

cmadam committed:
Save metadata after overwriting output directory

daw3rd committed:
make AbstractLauncherTest to allow spark to extend

Signed-off-by: David Wood <dawood@us.ibm.com>

cmadam committed:
Data used for tests and local runs

Signed-off-by: Constantin M Adam <cmadam@us.ibm.com>

cmadam committed:
Project utility files (copied from noop ray)

Signed-off-by: Constantin M Adam <cmadam@us.ibm.com>

cmadam committed:
Added spark config profiles for noop transform

daw3rd committed:
get spark noop_local.py running --help and build spark venv

Signed-off-by: David Wood <dawood@us.ibm.com>

daw3rd committed:
fill out spark's runtime __init__.py

Signed-off-by: David Wood <dawood@us.ibm.com>

daw3rd committed:
fill out spark's runtime __init__.py, the right one this time

Signed-off-by: David Wood <dawood@us.ibm.com>

cmadam committed:
Pass local config file path to the test

cmadam committed:
Merge branch 'spark-runtime' of github.com:IBM/data-prep-kit into spark-runtime

daw3rd committed:
change spark _key to _cli

Signed-off-by: David Wood <dawood@us.ibm.com>

daw3rd committed:
Merge branch 'spark-runtime' of github.com:IBM/data-prep-kit into spark-runtime

cmadam committed:
WIP: fix noop transformer

daw3rd committed:
fix spark __init__.py _key to _cli

Signed-off-by: David Wood <dawood@us.ibm.com>

daw3rd committed:
get noop spark Makefile working

Signed-off-by: David Wood <dawood@us.ibm.com>

cmadam committed:
Dockerfile that builds base pyspark image

cmadam committed:
Update the expected test data for spark noop

cmadam committed:
Implement validate directory for noop spark

daw3rd committed:
enable inclusion of runtime name in docker image

Signed-off-by: David Wood <dawood@us.ibm.com>

cmadam committed:
WIP: noop Dockerfile

cmadam committed:
Merge branch 'spark-runtime' of github.com:IBM/data-prep-kit into spark-runtime

daw3rd committed:
add make targets to build base lib images

Signed-off-by: David Wood <dawood@us.ibm.com>

cmadam committed:
Fixed typos introduced in last commit

Signed-off-by: Constantin M Adam <cmadam@us.ibm.com>

daw3rd committed:
Merge branch 'dev' into spark-runtime

daw3rd committed:
fix typos in makefile and rename multi_launcher test to ray_...

Signed-off-by: David Wood <dawood@us.ibm.com>

daw3rd committed:
fix type in name of data-processing-lib in recent commit to build-actions workflow

Signed-off-by: David Wood <dawood@us.ibm.com>

daw3rd committed:
more workflow/action changes for spark

Signed-off-by: David Wood <dawood@us.ibm.com>

daw3rd committed:
get make test target to pass in spark lib

Signed-off-by: David Wood <dawood@us.ibm.com>

blublinsky commented on data-processing-lib/ray/src/data_processing/transform/abstract_transform.py:
```diff
@@ -0,0 +1,16 @@
+        :return: a tuple of a list of 0 or more converted tables and a dictionary of statistics that will be
+        propagated to metadata
+        """
+        raise NotImplemented()
```
Why do we need this additional indirection?
Unlike Java Python does not distinguish methods based on the signature. So one more method of indirection creates more issues in the long run. What is the advantage?

daw3rd replied:
It formalizes the notion of a "transform" which we can then use across runtimes.  I'd like to have some commonality across them and not look like we're reinventing.

blublinsky replied:
This is a very noble intention, but do we really need this? Can you name at least one realistic use case where this can be used? At the moment, it's an extra baggage that does not add anything

daw3rd replied:
from below
`This comes in because of the AbstractTransformLauncher -> runtime config -> transform config -> AbstractTransform. Spark code is using the AbstractTransformLauncher and therefore we have a common AbstractTransform across all runtimes/launchers.`

blublinsky replied:
And how is this helping us exactly?

blublinsky commented on data-processing-lib/ray/src/data_processing/transform/binary_transform.py:
```diff
@@ -10,10 +10,15 @@
+
+DATA = TypeVar("DATA")
+
+
```
This class defines different method names. No need to define data

daw3rd replied:
But we need to enable passing DATA to the super class as a generic type.  See line 21 below.

blublinsky replied:
That is the complexity I am talking about. Why?

daw3rd replied:
from comment below
`This comes in because of the AbstractTransformLauncher -> runtime config -> transform config -> AbstractTransform. Spark code is using the AbstractTransformLauncher and therefore we have a common AbstractTransform across all runtimes/launchers.`

blublinsky replied:
And the same question. How is this helping us?

blublinsky commented on data-processing-lib/ray/src/data_processing/transform/table_transform.py:
```diff
@@ -59,17 +59,19 @@ def transform_binary(self, byte_array: bytes, ext: str) -> tuple[list[tuple[byte
+    #     propagated to metadata
+    #     """
+    #     raise NotImplemented()
 
```
This makes it less readable

daw3rd replied:
We're using the generic defintion of transform() from the super class so it is not needed here.

blublinsky replied:
Again, we sacrifice the readability of the code, and what is the gain? A simple cost/benefit analysis

daw3rd replied:
Copying from below
`This comes in because of the AbstractTransformLauncher -> runtime config -> transform config -> AbstractTransform. Spark code is using the AbstractTransformLauncher and therefore we have a common AbstractTransform across all runtimes/launchers.`

blublinsky replied:
And the same answer as above

daw3rd replied:
We have a common/unified framework, not independent ones.

blublinsky commented on data-processing-lib/ray/test/data_processing_tests/launch/pure_python/multi_launcher_test.py:
```diff
@@ -51,6 +51,10 @@ def _submit_for_execution(self) -> int:
+import pytest
+
+
+@pytest.mark.skip(reason="failing for now aws credentials")
```
please fix. It has to be something on your box. Its passing everywhere else

daw3rd replied:
Thank you and fixed,  and yes this passes at the command line and probably would in pycharm if it would use my existing env vars :(.

blublinsky commented on data-processing-lib/ray/test/data_processing_tests/launch/ray/ray_multi_launcher_test.py:
```diff
@@ -54,6 +54,10 @@ def _submit_for_execution(self) -> int:
+import pytest
+
+
+@pytest.mark.skip(reason="failing for now aws credentials")
```
Same here

daw3rd replied:
also fixed

blublinsky commented on data-processing-lib/spark/Dockerfile:
```diff
@@ -0,0 +1,89 @@
+USER ${spark_uid}
+
+# add pyspark, py4j to PYTHONPATH
+ENV PYTHONPATH=${SPARK_HOME}/python/lib/pyspark.zip:${SPARK_HOME}/python/lib/py4j-0.10.9.7-src.zip:${PYTHONPATH}
```
How is this better compared to prebuild PySpark image - https://hub.docker.com/layers/bitnami/spark/3.5.1/images/sha256-88b317b10433b9d87bbc409460716ef006f67f29433e89d13f3e7a8ac170d701?context=explore? The build is here https://github.com/bitnami/containers/tree/main/bitnami/spark/3.5/debian-12. In general, I think the common rule should be to avoid creating our own images, unless we absolutely have to


Alternatively Spark provides https://github.com/apache/spark/blob/master/bin/docker-image-tool.sh, which can be used to build image with this dockerfile https://github.com/apache/spark/blob/master/resource-managers/kubernetes/docker/src/main/dockerfiles/spark/bindings/python/Dockerfile

https://atirek-ak.medium.com/building-an-apache-spark-image-from-spark-binaries-fe1adeb9a964 has step by step instructions

cmadam replied:
1. We are already using `docker-image-tool.sh` tool to build the image.  Using this script directly is not a good practice, as it would require downloading a spark release on the local machine, and then running this script.  This approach cannot be automated, so we have encapsulated the download of the spark release and the invocation of this script in the docker image.

2. The bitnami spark image has several shortcomings - it is not built for running Spark in kubernetes, and it is not built for running Spark with S3.  As such, it requires several changes made to it, for example changing the `ENTRYPOINT`, without which it will not run in Kubernetes (on the driver node, or on the executor nodes). So we cannot use this image as is, and at this point I do not see any value in starting from this image, rather than from our base image.  Moreover, by looking at the size of this image (1.65GB), it is much larger than the size of our image (945MB); to use this image we will need to do additional investigation as to which packages are included in it, and where does the difference in size come from.

3. The article you cited contains an approach that has the drawback I listed under point 1 - it assumes that we are manually downloading a spark release on our machine and then we are manually running the `docker-image-tool.sh` script

blublinsky commented on data-processing-lib/spark/src/data_processing_spark/runtime/spark/__init__.py:
```diff
@@ -0,0 +1,8 @@
+from data_processing_spark.runtime.spark.runtime_config import SparkTransformRuntimeConfiguration
```
Can we, please, remove empty init files above?

blublinsky commented on data-processing-lib/spark/Dockerfile:
```diff
@@ -0,0 +1,89 @@
+# ARG PYTHON_IMAGE=python:3.11.9-slim-bookworm
+# FROM $PYTHON_IMAGE
+FROM python:3.11.9-slim-bookworm
+
```
We need to support Python 3.10 as well. See here https://medium.com/programmers-journey/deadsimple-pyspark-docker-spark-cluster-on-your-laptop-9f12e915ecf4 about Python versions.

blublinsky commented on data-processing-lib/spark/README.md:
```diff
@@ -0,0 +1,4 @@
+# Spark Transform Runtime 
+
+This is the extension of the core data-prep-kit library to support running transform
+in spark cluster.
```
Thats it? Probably need some more

daw3rd replied:
for sure, but I don't want to hold up the code changes for documentation

blublinsky commented on data-processing-lib/spark/config/spark_profile_kube.yml:
```diff
@@ -0,0 +1,14 @@
+# configuration needed when running in kubernetes
+spark.kubernetes.authenticate.driver.serviceAccountName: ${SERVICE_ACCOUNT}
+spark.kubernetes.container.image: ${EXECUTOR_DOCKER_IMAGE}
+spark.kubernetes.namespace: ${EXECUTOR_NAMESPACE}
```
I am not convinced we need this. When running Spark on Kubernetes, see https://blog.devgenius.io/spark-operator-basics-2ad7ecd469e1 for operator and why approach

blublinsky commented on data-processing-lib/spark/src/data_processing_spark/runtime/spark/spark_execution_config.py:
```diff
@@ -0,0 +1,45 @@
+            f"--{local_config_path_cli}",
+            type=str,
+            default=local_config_path_default,
+            help="Path to spark configuration for run",
```
I think this is bad. Spark config has to be defined outside an application itself. See https://github.com/kubeflow/spark-operator/blob/master/docs/user-guide.md for the example. And we need a "standard config for local run"

blublinsky commented on data-processing-lib/spark/src/data_processing_spark/runtime/spark/spark_execution_config.py:
```diff
@@ -0,0 +1,45 @@
+            type=str,
+            default=kube_config_path_default,
+            help="Path to Kubernetes-based configuration.",
+        )
```
I think this is quite confusing for the user. I think the better option is to just define local vs remote run

blublinsky commented on data-processing-lib/spark/src/data_processing_spark/runtime/spark/spark_launcher.py:
```diff
@@ -0,0 +1,219 @@
+
+            server_port = os.environ["KUBERNETES_SERVICE_PORT"]
+            master_url = f"k8s://https://kubernetes.default:{server_port}"
+
```
See comment above. I do not think it is the right approach. We should bite the bullet and adopt external configuration https://github.com/kubeflow/spark-operator/blob/master/docs/user-guide.md. This will make code smaller and provide a better configurability moving forward

daw3rd replied:
Your point here is a valid one and we should pursue it after we get this initial implementation in.  Is there any reason why we should not have this initial implementation, even if we decide to deprecate it in the future?  Would it somehow make all the class/launcher design obsolete? If yes, can you say why?

blublinsky replied:
There is nothing more permanent as temporary solution

blublinsky commented on data-processing-lib/spark/src/data_processing_spark/runtime/spark/spark_transform.py:
```diff
@@ -0,0 +1,12 @@
+        self.config = config
+
+    def transform(self, data: DataFrame) -> tuple[list[DataFrame], dict[str, Any]]:
+        raise NotImplemented()
```
This is fundamentally different from Python/Ray. Just bite a bullet and create a separate AbstractSparkTransform. Adding here AbstractTransform does not really add anything to the code. Its OO for the sake of OO nothing else. Also by using DataFrames here we are effectively limiting input to the Arrow tables. Just making sure that we understand this. There is no way back to the binary blobs

daw3rd replied:
This comes in because of the AbstractTransformLauncher -> runtime config -> transform config -> AbstractTransform.  Spark code is using the AbstractTransformLauncher and therefore we have a common AbstractTransform across all runtimes/launchers.

blublinsky replied:
So again, we are making this more complex and less readable and for what?

daw3rd replied:
complexity is in the ey of the beholder :)

blublinsky commented on data-processing-lib/spark/src/data_processing_spark/test_support/launch/abstract_launcher_test.py:
```diff
@@ -0,0 +1,8 @@
+from data_processing.test_support.launch.transform_test import (
```
Please remove empty init files

daw3rd replied:
Done.

blublinsky commented on data-processing-lib/spark/src/data_processing_spark/runtime/spark/spark_launcher.py:
```diff
@@ -0,0 +1,219 @@
+
+
+class SparkTransformLauncher(AbstractTransformLauncher):
+    """
```
Sorry this is not launcher. This is a combination of launcher and executor. Please split this into 2 classes:
1. Something that in Ray is called orchestrator
2. Actual launcher

daw3rd replied:
orchestrator is an implementation detail and the "orchestration" is essentially done in side _run_transform() with  spark implementing the "orchestration" piece.  if useful, we can certainly add a separate orchestrator later.

daw3rd replied:
and actually, i think the notion of "orchestration" is a ray-specific implementation detail.  That is, our ray implementation has ray worker that is performing the orchestration - very ray specific IMHO.  I don't see why we need to  have an "orchestrator" for python for example?

blublinsky replied:
It is not. The responsibility of the launcher is to launch, not define the execution. Execution definition (orchestration) is a separate concern from launching

daw3rd replied:
i respectfully disagree

blublinsky commented:
This is a good starting point, which I think needs quite a few changes. Another big question for me is that I do not see how this can be extended to support dedups or even bloclisting

daw3rd committed:
remove pytest ignores on multi launcher tests

Signed-off-by: David Wood <dawood@us.ibm.com>

cmadam committed:
lib noop test execution

Signed-off-by: Constantin M Adam <cmadam@us.ibm.com>

cmadam committed:
Merge branch 'spark-runtime' of github.com:IBM/data-prep-kit into spark-runtime

daw3rd committed:
get spark lib test passing on noop launch

Signed-off-by: David Wood <dawood@us.ibm.com>

daw3rd committed:
get get noop transform source test passing

Signed-off-by: David Wood <dawood@us.ibm.com>

daw3rd committed:
set spark noop image version

Signed-off-by: David Wood <dawood@us.ibm.com>

cmadam committed:
Change s3:// to s3a:// when reading multiple s3 files

daw3rd committed:
address empty __init__.py files in spark lib

Signed-off-by: David Wood <dawood@us.ibm.com>

cmadam committed:
Merge branch 'spark-runtime' of github.com:IBM/data-prep-kit into spark-runtime

cmadam committed:
Fixed isinstance list typo

Signed-off-by: Constantin M Adam <cmadam@us.ibm.com>

daw3rd committed:
extend spark launcher to allow AWS env vars w/ or w/o _IN or _OUT suffixes

Signed-off-by: David Wood <dawood@us.ibm.com>

cmadam committed:
Fixed isinstance list typo

Signed-off-by: Constantin M Adam <cmadam@us.ibm.com>

daw3rd committed:
Merge branch 'spark-runtime' of github.com:IBM/data-prep-kit into spark-runtime

daw3rd committed:
stub out _validate_metadata_content() in absract spark luancher test

Signed-off-by: David Wood <dawood@us.ibm.com>

daw3rd committed:
make spark read s3 credentials from DataAccessS3

Signed-off-by: David Wood <dawood@us.ibm.com>

cmadam committed:
Update Noop and Filter Spark image versions

Signed-off-by: Constantin M Adam <cmadam@us.ibm.com>

cmadam committed:
Spark filter transform

Signed-off-by: Constantin M Adam <cmadam@us.ibm.com>

cmadam committed:
Add the transform version for Spark document ID

Signed-off-by: Constantin M Adam <cmadam@us.ibm.com>

cmadam committed:
Always stop the spark session at the end of transform

Signed-off-by: Constantin M Adam <cmadam@us.ibm.com>

cmadam committed:
Spark generate document ID transform

Signed-off-by: Constantin M Adam <cmadam@us.ibm.com>

cmadam committed:
Added Spark data processing lib documentation

Signed-off-by: Constantin M Adam <cmadam@us.ibm.com>

cmadam committed:
Update doc for filter transform

Signed-off-by: Constantin M Adam <cmadam@us.ibm.com>

cmadam committed:
Add Spark launcher options doc

Signed-off-by: Constantin M Adam <cmadam@us.ibm.com>

daw3rd committed:
add debug 'docker images' to universal build action

Signed-off-by: David Wood <dawood@us.ibm.com>

daw3rd committed:
Merge branch 'dev' into spark-runtime

daw3rd committed:
Correct title for tools image build action

Signed-off-by: David Wood <dawood@us.ibm.com>

daw3rd committed:
Merge branch 'spark-runtime' of github.com:IBM/data-prep-kit into spark-runtime

Signed-off-by: David Wood <dawood@us.ibm.com>

cmadam committed:
Kubernetes launch is WIP, disabling for now

Signed-off-by: Constantin M Adam <cmadam@us.ibm.com>

daw3rd committed:
Fix spark transform image build.

Support base name specification in image builds.
Also split DOCKER_IMAGE into DOCKER_LOCAL/REMOTE_IMAGE to fix spark image build so it can find the local image.

Signed-off-by: David Wood <dawood@us.ibm.com>

daw3rd committed:
Merge branch 'spark-runtime' of github.com:IBM/data-prep-kit into spark-runtime

daw3rd committed:
Fix ingest2parquet image build and rename DOCKER_NAME to DOCKER_IMAGE_NAME in makefiles

Signed-off-by: David Wood <dawood@us.ibm.com>

daw3rd committed:
Fix building of base spark image and set DOCKER_REMOTE_IMAGE using DOCKER_REGISTRY_ENDPOINT

Signed-off-by: David Wood <dawood@us.ibm.com>

daw3rd committed:
Change/fix kfkp_ray_components makefile to use DOCKER_LOCAL_IMAGE instead of hosted image

Signed-off-by: David Wood <dawood@us.ibm.com>

daw3rd committed:
Minor updates on spark-runtime doc.

Signed-off-by: David Wood <dawood@us.ibm.com>
